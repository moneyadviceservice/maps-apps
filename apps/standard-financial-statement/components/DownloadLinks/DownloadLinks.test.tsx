import { cleanup, fireEvent, render, screen } from '@testing-library/react';

import { DownloadLinks } from '.';

import '@testing-library/jest-dom';

// Mock useAnalytics hook
const mockAddEvent = jest.fn();
jest.mock('@maps-react/hooks/useAnalytics', () => ({
  useAnalytics: () => ({
    addEvent: mockAddEvent,
  }),
}));

const mockWindowOpen = jest.fn();
Object.defineProperty(window, 'open', {
  value: mockWindowOpen,
  writable: true,
});

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({
    query: {},
  }),
}));

const mockDownload = {
  fileName: 'Download the SFS Code of Conduct',
  asset: {
    size: 83064,
    _path:
      '/content/dam/sfs/assets/sfs-code-of-conduct/MAS0126_SFS_Code_of_Conduct_A_v1.pdf',
    type: 'document',
  },
};

describe('DownloadLinks component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  it('renders single download correctly', () => {
    render(<DownloadLinks assetPath="test" downloads={[mockDownload]} />);
    const downloads = screen.getByTestId('downloads');
    const fileSize = screen.getByTestId('file');
    const heading = screen.getByTestId('downloads-heading');
    expect(heading).toHaveTextContent('Download file');
    expect(downloads).toMatchSnapshot();
    expect(fileSize).toHaveTextContent('81 KB');
  });

  it('renders multiple downloads correctly', () => {
    render(
      <DownloadLinks
        assetPath="test"
        downloads={[mockDownload, mockDownload]}
      />,
    );
    const downloads = screen.getByTestId('downloads');
    const heading = screen.getByTestId('downloads-heading');
    expect(heading).toHaveTextContent('Download files');
    expect(downloads).toMatchSnapshot();
  });

  describe('Download analytics', () => {
    it('triggers analytics event when download link is clicked', () => {
      render(<DownloadLinks assetPath="test" downloads={[mockDownload]} />);

      // Find and click the download link
      const downloadLink = screen.getByText('Download the SFS Code of Conduct');
      fireEvent.click(downloadLink);

      // Verify analytics event was triggered with correct payload
      expect(mockAddEvent).toHaveBeenCalledWith({
        event: 'fileDownload',
        eventInfo: {
          linkText: 'Download the SFS Code of Conduct',
        },
      });

      // Verify window.open was called with correct URL
      expect(mockWindowOpen).toHaveBeenCalledWith(
        'http://localhost/test/content/dam/sfs/assets/sfs-code-of-conduct/MAS0126_SFS_Code_of_Conduct_A_v1.pdf',
        '_blank',
      );
    });

    it('prevents default behavior when download link is clicked', () => {
      render(<DownloadLinks assetPath="test" downloads={[mockDownload]} />);

      const downloadLink = screen.getByText('Download the SFS Code of Conduct');
      const mockEvent = {
        preventDefault: jest.fn(),
        currentTarget: {
          textContent: 'Download the SFS Code of Conduct',
          href: 'http://localhost/test/content/dam/sfs/assets/sfs-code-of-conduct/MAS0126_SFS_Code_of_Conduct_A_v1.pdf',
        },
      };

      // Simulate the click event with our mock
      const clickEvent = new MouseEvent('click', { bubbles: true });
      Object.defineProperty(clickEvent, 'preventDefault', {
        value: mockEvent.preventDefault,
      });
      Object.defineProperty(clickEvent, 'currentTarget', {
        value: mockEvent.currentTarget,
      });

      downloadLink.dispatchEvent(clickEvent);

      // Verify preventDefault was called
      expect(mockEvent.preventDefault).toHaveBeenCalled();
    });

    it('handles download with different file types correctly', () => {
      const docxDownload = {
        fileName: 'Test Document',
        asset: {
          size: 1024,
          _path: '/content/dam/sfs/assets/test-document.docx',
          type: 'document',
        },
      };

      render(<DownloadLinks assetPath="test" downloads={[docxDownload]} />);

      const downloadLink = screen.getByText('Test Document');
      fireEvent.click(downloadLink);

      // Verify analytics event was triggered with correct file type
      expect(mockAddEvent).toHaveBeenCalledWith({
        event: 'fileDownload',
        eventInfo: {
          linkText: 'Test Document',
        },
      });

      // Verify window.open was called with correct URL
      expect(mockWindowOpen).toHaveBeenCalledWith(
        'http://localhost/test/content/dam/sfs/assets/test-document.docx',
        '_blank',
      );
    });
  });
});
