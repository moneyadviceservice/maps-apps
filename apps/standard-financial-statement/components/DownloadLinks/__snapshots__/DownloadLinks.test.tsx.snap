// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadLinks component renders multiple downloads correctly 1`] = `
<div
  class="mt-10"
  data-testid="downloads"
>
  <h4
    class="text-xl md:text-2xl font-bold mb-6 text-blue-600"
    data-testid="downloads-heading"
  >
    <span>
      Download files 
    </span>
  </h4>
  <div
    class="flex items-center mb-7"
  >
    <a
      class="inline-flex items-center visited:text-purple-700 focus:text-gray-800 focus:bg-yellow-400 focus:shadow-link-focus focus-within:outline-none underline hover:no-underline focus:no-underline cursor-pointer active:text-gray-800 active:underline active:bg-transparent active:shadow-none t-link-button gap-0 px-0 mx-0 text-magenta-800 text-[18px] font-bold"
      href="test/content/dam/sfs/assets/sfs-code-of-conduct/MAS0126_SFS_Code_of_Conduct_A_v1.pdf"
      target="_blank"
    >
      <span />
      Download the SFS Code of Conduct
      <span>
        <download.svg
          class="ml-2"
        />
      </span>
    </a>
    <div
      class="ml-2 text-gray-500"
    >
      <span
        class="uppercase"
      >
        (
        pdf
        ,
      </span>
       
      <span
        data-testid="file"
      >
        81 KB
        )
      </span>
    </div>
  </div>
  <div
    class="flex items-center mb-7"
  >
    <a
      class="inline-flex items-center visited:text-purple-700 focus:text-gray-800 focus:bg-yellow-400 focus:shadow-link-focus focus-within:outline-none underline hover:no-underline focus:no-underline cursor-pointer active:text-gray-800 active:underline active:bg-transparent active:shadow-none t-link-button gap-0 px-0 mx-0 text-magenta-800 text-[18px] font-bold"
      href="test/content/dam/sfs/assets/sfs-code-of-conduct/MAS0126_SFS_Code_of_Conduct_A_v1.pdf"
      target="_blank"
    >
      <span />
      Download the SFS Code of Conduct
      <span>
        <download.svg
          class="ml-2"
        />
      </span>
    </a>
    <div
      class="ml-2 text-gray-500"
    >
      <span
        class="uppercase"
      >
        (
        pdf
        ,
      </span>
       
      <span
        data-testid="file"
      >
        81 KB
        )
      </span>
    </div>
  </div>
</div>
`;

exports[`DownloadLinks component renders single download correctly 1`] = `
<div
  class="mt-10"
  data-testid="downloads"
>
  <h4
    class="text-xl md:text-2xl font-bold mb-6 text-blue-600"
    data-testid="downloads-heading"
  >
    <span>
      Download file
    </span>
  </h4>
  <div
    class="flex items-center mb-7"
  >
    <a
      class="inline-flex items-center visited:text-purple-700 focus:text-gray-800 focus:bg-yellow-400 focus:shadow-link-focus focus-within:outline-none underline hover:no-underline focus:no-underline cursor-pointer active:text-gray-800 active:underline active:bg-transparent active:shadow-none t-link-button gap-0 px-0 mx-0 text-magenta-800 text-[18px] font-bold"
      href="test/content/dam/sfs/assets/sfs-code-of-conduct/MAS0126_SFS_Code_of_Conduct_A_v1.pdf"
      target="_blank"
    >
      <span />
      Download the SFS Code of Conduct
      <span>
        <download.svg
          class="ml-2"
        />
      </span>
    </a>
    <div
      class="ml-2 text-gray-500"
    >
      <span
        class="uppercase"
      >
        (
        pdf
        ,
      </span>
       
      <span
        data-testid="file"
      >
        81 KB
        )
      </span>
    </div>
  </div>
</div>
`;
